### DISCUSSION TITLE
Refactoring

### DISCUSSION THREAD

**Discussion by TornaxO7** [leftwm/leftwm#865] [Ideas]
## Motivation
Some files/functions are too big and (a little bit) complicated.
To name some:
    - [`window_handler.rs`](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/handlers/window_handler.rs) has basically four important function which can be separated into for files: `window_handler/create.rs`, `window_handler/destroy.rs`, `window_handler/change.rs` for example.
    - the [`config trait`](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/config/mod.rs#L13) has more functions than it should like [`save_state`](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/config/mod.rs#L59). A config should represent the config file and should not also do some event-loop actions. Those are two usages which can be separated
    - [`window`](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/models/window.rs#L42) is probably a little bit too big (has a lot of attributes) and represents a tiled but also a floating window. This could be probably separated as well into a `TiledWindow` and `FloatingWindow` for example with a `Window` trait.
 
etc.

## Proposal
I'd propose the following roadmap:

1. Create a discussion thread where all users can write their use-cases down,
    like what they're expecting to have and how it should work
2. Collect the main parts, what `leftwm` should have in the end like:
    - workspaces
    - tags
    - layouts
    - "plugin"/"exension"-system (for `leftwm-theme`, `leftwm-config` etc.)
    - etc.
4. Discuss the relations between them.
5. Discuss how they should be implemented
6. Start refactoring

### Pros
    - Stable running version
    - clean code base (hopefully lol)
    - documented code (hopefully lol)

### Cons
    - risc of losing attention => a lot of work would be lost
    - takes time (a lot of) time => no bugfixes and new features of new issues possible

## Question
What do you think about that proposal? @AethanFoot @VuiMuich @hertg 

## Other information
If you agree a general-big-refactoring, please give this a thumb up and give this a dislike if you disagree.
I'll wait for the result until next sunday (09.10.2022 [DD/MM/YYYY]). (@VuiMuich @AethanFoot @hertg I think your votes are the most important since you are the maintainers :))

**Comment by AethanFoot** [comment]
The only thing I disagree with is the window bit, windows have a lot of properties, thats just a thing (there are some things that are the same across all windows, but would be less ergonomic to access from else where). And splitting into 2 types is very object oriented, and is a thing frowned upon by rust devs (inheritance and classes arent a thing for a reason, and rust isn't object oriented).

Other than that I would agree but with a few factors; the core dev team is quite small so things will take time (this is why we have been slowly refactoring as we go), we have to keep in mind this isnt a formal settings so organising things like this can be challenging (and this is a passion project for most), we would have to ignore new PR's from external devs/would have to be very strict and critical of PR's handed in (which can be discouraging for new devs).  

Once things are planned we have to remember since it is open source, the strategy has to be extremely agile (ideas will constantly change and evolve, yes we may have a plan but it needs to be flexible). We will need to make the ideas clear and easy to find (e.g. make issues for them as this is accessible for new and inexperienced devs), for project boards may be nice but unfamiliar and not accessible to all users, this is important so other users have a chance to contribute.

In my mind, changing the trait to configurator solves most of the confusion and the config trait (but i agree certain things could be moved), in my head the config struct is created by the user and used to prime the configurator for leftwm-core to use.

Overall sounds good and would be the step needed to take leftwm out of "beta" (a 1.0.0 release). But we have to remember that the environment isn't so strict and even in "real" work it is quite flexible.

**Reply by AethanFoot** [reply]
I will add, we could break down the window into different structs that store groups of properties to make it clearer, e.g.
```
pub struct Window {
    pub x_properties: XProperties,
    ...
}

**Reply by TornaxO7** [reply]
> The only thing I disagree with is the window bit, windows have a lot of properties, thats just a thing (there are some things that are the same across all windows, but would be less ergonomic to access from else where). And splitting into 2 types is very object oriented, and is a thing frowned upon by rust devs (inheritance and classes arent a thing for a reason, and rust isn't object oriented).

Hm... I was thinking about something like that:
```rust
pub trait Window {
    fn redraw(&self);
}

/// data which is relevant for *all* windows
pub struct WindowData {
    margin: ...,
    border: ...,
}

pub struct TiledWindow {
    data: WindowData,
    // data only relevant to tiled
}

pub struct FloatingWindow {
    data: WindowData,
    // data only relevant to floating
}

impl Window for TiledWindow {
    // ...
}

impl Window for FloatingWindow {
    // ...
}

/// conversion from tiled to floating:
/// usage would be like: `let floating = FloatingWindow::from(window) // window: TiledWindow`
impl From<TiledWindow> for FloatingWindow {
    fn from(tiled: TiledWindow) -> Self {
        // ...
    } 
}
```
this is just an example! Attributes might be wrong and the struct would be in different files! I think we're getting a bit too much into details xD The origin of my discussion here is just, if we should start this refactoring according to my roadmap. xD

**Reply by AethanFoot** [reply]
I cant think of anything a TiledWindow would have that FloatingWindow wouldn't, also sorting and iteratoring through a vector with 2 different types would add more complexity to anything manipulating the windows vec

**Reply by VuiMuich** [reply]
I have an idea, why the `save_state` and `load_state` were put there, but I agree, that they should be moved. The reasoning probably was that loading the previous state is kind of a configuring operation, and saving should live along loading.

The rest of the functions belong right there I guess, but some name refactoring should improve further to understand what the trait is meant to do,  as they just use the config field names, but functions should use verb based names.

**Reply by TornaxO7** [reply]
> I cant think of anything a TiledWindow would have that FloatingWindow wouldn't, also sorting and iteratoring through a vector with 2 different types would add more complexity to anything manipulating the windows vec

Ok, you might be right here, I can't collect the points in my mind as well at the moment.

**Comment by Eskaan** [comment]
I would like to put the display server implementation into it's own internal crate to improve (development) compile times, so if you are doing something outside of the server implementation it won't have to recompile every time.
I expect this to be little to no effort as its already separated by a trait.

**Reply by AethanFoot** [reply]
Yeah this is something I would like to work towards, as yes there will be more packages to manage (but many would remain mostly static). This as you said would help compile times (something I hadn't actually thought about), but will add to the "build your own window manager" we would like to work towards (so they could in theory choose whatever display server they want).

**Reply by TornaxO7** [reply]
>  would like to put the display server implementation into it's own internal crate to improve (development) compile times,

May I ask how it improve development compile times? Isn't rust usign incremental compilation? So it will only compile your stuff which has changed or am I wrong?

**Reply by Eskaan** [reply]
Exactly, so if you don't change the dispay_server crate it only has to compile the now smaller leftwm-core (...).
It improves compile times because leftwm-core is smaller.

**Reply by TornaxO7** [reply]
but it doesn't compile the `display_server` crate anyhow if we don't change anything there or am I missunderstanding something here?

**Reply by AethanFoot** [reply]
Because currently the display server code within leftwm-core gets recompiled whenever something changes (as it isn't separated)

**Reply by Eskaan** [reply]
Part one (refactor): #895
Part two will be an implementation to select which of the compiled display servers to launch.

**Comment by VuiMuich** [comment]
Three things I would cosider refactor worthy appart topics mentioned above:
- reduce places that need to be edited for changes to/adding new commands and derive command list in `leftwm command --list`
- collect any file handling (config and theme read, write defaults, migrate etc.) into a module — maybe even `save_state` could move there
- collect `tests` either to a module or a crate

**Comment by TornaxO7** [comment]
I'd say that I'll wait for more readers and look at this again after... a week (until the sunday next week (inclusive))? Please give my first post a thumbs-up if you agree for a general-big-refactoring!

**Comment by Antikyth** [comment]
I am not necessarily _suggesting_ or _recommending_ here, but considering this is a big refactor, I think it might just be worth mentioning the projects I have been working on and have planned outside of LeftWM.

I've been working on quite a lot since June this year involving X, particularly in Rust. Specifically, I have been designing and discussing the design of these projects with others:
 - A direct implementation of the X11 protocol formats and data structures in Rust. The purpose of this is to serve as a foundation upon which 'API wrappers' can be developed; i.e. X libraries that provide an opinionated and hand-crafted API for X by providing wrappers around these 1:1 representations of X. This project is called _X Rust Bindings_, or XRB for short. Currently I expect it to be near feature complete at the end of the year.
- An X library built for (async) Rust based on XRB, called _X.RS_. The goals of this library are:
  - to provide an intuitive and idiomatic X API catered towards Rust;
  - to use async Rust to make asynchronous code a lot easier and nicer to write when working with X, which, while possible with existing X libraries, should make asynchronous code much more widespread and should therefore improve the performance of software using X.RS quite a lot over software using existing X libraries;
  - and to provide an X library that has extensive documentation and comments, allowing it and X in general to be easily learnt.
- A custom X protocol that allows for further modularity and separation of the behaviours of window managers. This might be particularly interesting to LeftWM in terms of plugins/extensions. For example, window decorations, in my opinion, shouldn't necessarily be the responsibility of the window manager. This protocol could allow external 'decorator clients' to provide the window decorations for any complaint window manager, which would mean users could theme their window manager's decorations in the same way irrespective of the window manager being used, and any compliant window manager would be able to easily implement far more potential for customisation.

I mention these projects because I think they could either provide inspiration for the design of LeftWM, or they could even be used directly in LeftWM itself when complete. Currently, with just me working on them, they are quite far from completion: right now X.RS and the protocol specification haven't been started, and XRB has a large amount done, but still has loose ends to finish before it can be ready (hence my estimate of the end of this year). I don't know what would be most useful to LeftWM: you use them as inspiration, contribute to them to complete them quicker, or simply don't use them at all. I just wanted to share them considering this proposal for a large refactor of LeftWM - using any of these projects for LeftWM would require a large refactor anyway.

**Comment by TornaxO7** [comment]
Ok, since people are adding ideas and are liking/upvoting this discussion, I'll assume that we can start doing the refactoring by creating issues or discussions about a given point *what* needs to be refactored and how we can do it.

**Reply by VuiMuich** [reply]
I think starting issues for each of the mentioned refactors is a good start to track the idea, and further discuss the reach, approaches, possible collisions etc.
Also makes it easier to hint on it in upcomming PRs which might in some way be related, obsolete, colliding or what ever.

As a general tendency we would like the bigger refactors stand a bit back in favour of smaller bugfix PRs for the moment, as the latter might be more attractive for new contributors and Hacktoberfest participants.

**Comment by Eskaan** [comment]
Another suggestion just came to my mind. How about we unify the theme and the main config.

My idea is to make the theme config technically the same, but overriding the main config when options are set.
Currently, we take the default config, then override it with the main config. We could spare much complexity and improve theme customizability.
And the argument I think I can get VuiMuch with is, that he can just make a seperate theme using all the options in the weirdest configurations to test (my :sweat_smile:) new code.

**Reply by VuiMuich** [reply]
Similar ideas have been around a couple of times and basically we don't dislike this. There is a few things to consider though:
- technically the default does not get overridden, but is fallen back to, if the loaded config fails or is missing, if I am not mistaken
- allowing themes to override fields of the config is imo a must opt-in from the config. I.e. the config can allow to change Tag names, but not the number of available tags, or the theme might add additional scratchpads (if not conflicting in the name) etc. But a theme must never be able to override keybinds, at best it might be allowed to add some, if they were unused so far.
- the other way round might be useful as well: e.g. someone with daltonism might want to set border colors to override the theme ones so they always will be able to distinguish focused and normal borders, despite the theme's choice.

On a side note: a feature I would love to add would be a `PartialReload` command and this is probably not terribly complicated to add, as the `LoadTheme` command basically is already doing it: it changes a specific set of fields of the config independently of the rest of the config. The only tricky part might be to make sure there wont be a gap between dropping the old config values and loading the new ones in which strange things might happen. For themes it is not critical if this doesn't go completelly smooth, but messing up the workspace setup or keybinds of a running session might end badly. Typically rusts fearless concurrency should provide all needed safety here, but we want to be extra certain.
I think the `UnloadTheme` was added once just in this spirit, to guarantee a clean switch of theme configs. It still makes sense, when the old theme uses optional fields the new loaded does not specify, but otherwise is not really necessary to successfully apply a theme.
This `PartialReload` could be used for a bunch of situations, like just reloading workspace config after a hotplug event on the graphics card output, or just reloading keybnds, without resetting all the rest of the WM. And maybe even allow to only fallback to defaults in failing parts of the config during a regular reload (this might require some further features of the `ron-serde` though, that allow the error better awareness where the parsing fails).

**Reply by Eskaan** [reply]
First, about the Overrides:
Reading this I would propose a `override.ron` defining order and more importantly what gets overriden.

To the second proposal for `PartialReload`:
I like the idea, and I think this would be even easier using the code from the `overrides.ron` (specifically the enum with string to variable matching) for single values, alternatively just providing a file as argument and just running the same functions used for applying the main config.

**Reply by VuiMuich** [reply]
> First, about the Overrides: Reading this I would propose a `override.ron` defining order and more importantly what gets overriden.

Sounds interesting

> To the second proposal for `PartialReload`: I like the idea, and I think this would be even easier using the code from the `overrides.ron` (specifically the enum with string to variable matching) for single values, alternatively just providing a file as argument and just running the same functions used for applying the main config.

Yeah, basically the `PartialReload` would be similar to the `-s` flag in `leftwm-state` allowing to pass a list of overrides as a string. Also a nice idea.

**Reply by Eskaan** [reply]
*Another reply to your first message in this thread*
We could take a Enum to either add-only, edit or replace-all. So if you want to replace all keybinds but only edit tag names you do
```rust
(config: "./file.ron", type: Replace, use: [keybind])
(config: "./file.ron", type: Edit, use: [tag])
```
This would allow for highly advanced, but simple configuration.

If a theme edits the workspaces or something else it shouln‘t I would just consider this bad practice as we do when themes don‘t use their own config files. But some may want to create his own custom themes with different keybinds.
